lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/TestS3Retries.java

```diff
@@ -0,0 +1,122 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem.s3;
+
+import eu.rekawek.toxiproxy.Proxy;
+import eu.rekawek.toxiproxy.ToxiproxyClient;
+import eu.rekawek.toxiproxy.model.ToxicDirection;
+import io.airlift.log.Level;
+import io.airlift.log.Logging;
+import io.trino.filesystem.Location;
+import io.trino.testing.containers.Minio;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+import org.testcontainers.containers.Network;
+import org.testcontainers.containers.ToxiproxyContainer;
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Arrays;
+
+import static io.trino.testing.containers.Minio.MINIO_API_PORT;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+public class TestS3Retries
+{
+    Network network;
+    private Minio minio;
+    private ToxiproxyContainer toxiproxy;
+
+    @BeforeAll
+    final void init()
+            throws IOException
+    {
+        Logging logging = Logging.initialize();
+        logging.setLevel("software.amazon.awssdk.request", Level.DEBUG);
+
+        network = Network.newNetwork();
+        minio = Minio.builder()
+                .withNetwork(network)
+                .build();
+        minio.start();
+        minio.createBucket("bucket");
+        minio.writeFile(getTestData(), "bucket", "object");
+
+        toxiproxy = new ToxiproxyContainer("ghcr.io/shopify/toxiproxy:2.5.0")
+                .withExposedPorts(8474, 1234)
+                .withNetwork(network)
+                .withNetworkAliases("minio");
+        toxiproxy.start();
+
+        ToxiproxyClient toxiproxyClient = new ToxiproxyClient(toxiproxy.getHost(), toxiproxy.getControlPort());
+        Proxy proxy = toxiproxyClient.createProxy("minio", "0.0.0.0:1234", "minio:" + MINIO_API_PORT);
+        // the number of bytes includes both the response headers (around 570 bytes) and body
+        proxy.toxics()
+                .limitData("broken connection", ToxicDirection.DOWNSTREAM, 700);
+    }
+
+    private static byte[] getTestData()
+    {
+        byte[] data = new byte[1024];
+        Arrays.fill(data, (byte) 1);
+        return data;
+    }
+
+    @AfterAll
+    final void cleanup()
+    {
+        minio.close();
```

Comment:
Do this using a `Closer`. You can add objects to it like
```java
network = closer.register(Network.newNetwork());
```
This ensures we close everything, even if one of the closes fails. It also does the closes in the reverse automatically.

Reply:
Much nices, thanks!